File: programmers/level2/캐시/javascript/20230113/solution.js
function toLowerCase(cities) {
  return cities.map(city => city.toLowerCase());
}

function solution(cacheSize, cities) {
  let runtime = 0;
  const citiesToLowerCase = toLowerCase(cities);
  let cacheList = [];

  for (const city of citiesToLowerCase) {
    if (cacheSize === 0) {
      runtime += 5;
      continue;
    }

    if (cacheList.length > cacheSize) {
      cacheList.shift();
    }

    if (cacheList.includes(city)) {
      runtime += 1;
      cacheList = cacheList.filter(item => item !== city);
      cacheList.push(city);
    }

    if (!cacheList.includes(city)) {
      runtime += 5;
      cacheList.push(city);
    }
  }

  return runtime;
}

File: programmers/level2/캐시/java/20230113/src/Solution.java
import java.util.LinkedList;
import java.util.Queue;

public class Solution {
    public int solution(int cacheSize, String[] cities) {
        int runtime = 0;
        String[] citiesToLowerCase = toLowerCase(cities);
        Queue<String> queue = new LinkedList<>();

        for (String city : citiesToLowerCase) {
            if (cacheSize == 0) {
                runtime += 5;
                continue;
            }

            if (queue.size() > cacheSize) {
                queue.remove();
            }

            if (queue.contains(city)) {
                runtime += 1;
                queue.remove(city);
                queue.add(city);
            }

            if (!queue.contains(city)) {
                runtime += 5;
                queue.add(city);
            }
        }

        return runtime;
    }

    public String[] toLowerCase(String[] cities) {
        String[] array = new String[cities.length];

        for (int i = 0; i < cities.length; i += 1) {
            array[i] = cities[i].toLowerCase();
        }

        return array;
    }
}
